Analysis: Add support for overriding variable/constant BPM on a per-track basis

fwcd opened the pull request:
This PR, intended as a follow-up to #4806, wraps scheduled tracks in `AnalyzerScheduledTrack` and `AnalyzerTrack`, depending on whether it contains a `TrackId` or a resolved `TrackPointer`.

These wrappers add the ability to override analysis options on a per-track basis. This PR also introduced one such option, namely `useFixedTempo`, which can be used to override the constant/variable BPM setting. Additionally, it adds two new entries to `WTrackMenu` > `Analyze`:

- Reanalyze (constant BPM)
- Reanalyze (variable BPM)

<img width="369" alt="image" src="https://user-images.githubusercontent.com/30873659/193075096-14a3cb43-6110-4171-849c-d848cb4f2f48.png">

Any feedback is of course welcome!

fwcd committed:
AnalyzerTrack: Add AnalyzerTrack

fwcd committed:
AnalyzerTrack: Use it instead of TrackPointers

fwcd committed:
AnalyzerScheduledTrack: Add AnalyzerScheduledTrack

fwcd committed:
AnalyzerScheduledTrack: Use it instead of track ids

fwcd committed:
WTrackMenu: Add ability to pass analysis options

fwcd committed:
AnalyzerBeats: Respect AnalyzerTrack::Options::useFixedTempo

fwcd committed:
WTrackMenu: Add fixed/variable bpm reanalysis actions

fwcd commented on src/analyzer/analyzerthread.h:
```diff
@@ -108,7 +110,7 @@ class AnalyzerThread : public WorkerThread {
     // NOTE(uklotzde, 2018-01-04): Ideally we would use std::atomic<TrackPointer>,
     // for this purpose, which will become available in C++20.
-    rigtorp::SPSCQueue<TrackPointer> m_nextTrack;
+    rigtorp::SPSCQueue<AnalyzerTrack> m_nextTrack;
```
Is this still thread-safe, even though we now use a wrapper structure here?

fwcd commented on src/analyzer/analyzerthread.cpp:
```diff
@@ -204,14 +204,14 @@ bool AnalyzerThread::submitNextTrack(TrackPointer nextTrack) {
+    AnalyzerTrack* pFront = m_nextTrack.front();
     if (pFront) {
-        m_currentTrack = *pFront;
+        *m_currentTrack = *pFront;
```
I'm getting segfaults here during the decrement of the `TrackPointer`'s reference count (inside the `AnalyzerTrack` that is getting replaced) and I'm not really sure why, since I don't see where the `AnalyzerTrack` is handled unsafely. Perhaps someone else has an idea what the issue could be?

fwcd committed:
AnalyzerTrack: Pass it by const reference

...to make clazy happy!

daschuer commented on src/analyzer/analyzerthread.h:
```diff
@@ -118,7 +120,7 @@ class AnalyzerThread : public WorkerThread {
     mixxx::SampleBuffer m_sampleBuffer;
 
-    TrackPointer m_currentTrack;
+    std::optional<AnalyzerTrack> m_currentTrack;
```
This is not a good fit for an optional, because the track inside inside can be still null even though the optional predicts that we have a value. You will have the same convenient API if you implement "has_value()" yourself without redundant data storage.

fwcd replied:
I would rather like to do it the other way around, i.e. assert that the `TrackPointer` inside must not be null, since to me it's semantically the right thing: If there's no track, we probably don't want to attach any options either, so it feels cleaner to me to model `AnalyzerTrack` as a value that is never empty and then add optionality externally.

daschuer replied:
Another question is if this can ever be null. We have recently added a std::not_null to our code base. 
I am afraid that this does not work with our smart track pointer. We need to have a look. 
In any case a not_null track pointer will help our code base a lot. (A topic for another PR ...)

fwcd replied:
Regarding

> without redundant data storage.

IMO it's not the right place to micro-optimize: The extra byte that `std::optional` uses would be a cost that we pay per `AnalyzerThread` and not a very significant one.

daschuer replied:
The discussion about the extra byte is mood, if we consider that the track is always valid. We can make this true, but lets do it in another PR to keep this one simple. 

Having two concurrent values the express an invalid state is a recipe of future issue that needs to be fixed IMHO. It is not that much a performance thing. For me the best solution for this PR is to add an own "has_value()" function.

fwcd replied:
Yeah, let's move that to a separate PR. But I think it makes sense _semantically_ that an analyzer thread could either have a track or not. If we were to model optionality inside of `AnalyzerTrack` we might want to consider renaming that class, since it no longer represents a 'track' but an optional track, I'd really vote against that thought. Keeping `AnalyzerTrack` an individual, non-nullable unit that can be scheduled is IMO cleaner and doesn't mix in another concern (optionality).

daschuer replied:
So we go for a "has_value()" member in this PR, right?

fwcd replied:
I would really advise against it, since it goes against the design of this class (being a POD value type), doesn't buy us anything in terms of thread-safety, simplicity and barely makes a difference in space/performance, but if you feel strongly about this, then we can of course also do that.

The reason I believe `std::optional` makes sense here is that we already effectively did the same thing already, just implicitly: We used null pointers instead of C++'s optionals. The latter, however, are simultaneously more clear, harder to misuse, encode invariants in the type system and really don't have a big footprint.

daschuer replied:
We need here a protecting against null pointer access. Using another value for it is errophrone.  

Look at our code base and the qt code base. The are many cases with an invalid state internally and a API function to access that. Wrapping that by an optional make the code harder to use and more complex. 

Optionals are a perfect fit for types like booleans without an invalid value. But please don't use it for pointers.

fwcd replied:
I have read up on the documentation of `gsl::not_null` and it seems to be supported to use it for smart pointers. So I've wrapped `AnalyzerTrack`'s `TrackPointer` in `gsl::not_null` and updated the tests as necessary. In `TrackAnalysisScheduler` (where we usually construct `AnalyzerTrack`s) we already perform a null-check, so constructing the `not_null` should be safe.

I believe this is the cleanest solution, since illegal states should now be unrepresentable and we can compose optionality separately on top (of a combined not-null track-with-options). It also improves local reasoning since optionality as an effect is more visible than having to worry about an internal pointer that may or may not be null.

Let me know if this works for you.

daschuer commented on src/widget/wtrackmenu.cpp:
```diff
@@ -1377,6 +1402,20 @@ void WTrackMenu::slotReanalyze() {
+void WTrackMenu::slotReanalyzeWithFixedTempo() {
+    clearBeats();
+    AnalyzerTrack::Options options;
+    options.useFixedTempo = true;
```
I think the API would be more future prove to have here also the option "from preferences" or such.

ronso0 commented:
Re track menu implementation:
Instead of 3 "Reanalyze" actions, how about reading the current setting (const/var BPM analyzer) and display these actions
* Analyze
* Reanalyze
* Reanalyze with [constant | variable] BPM

where [constant | variable] is the opposite of the current setting.

daschuer commented:
Yes this would help to get rid of the redundant menu. The other issue is that it is not that clear what exactly is reanalyzed with "Reanalyze" when we have two other options where BPM is explicit mentioned. 

Some ideas for discussion 

* Analyze
* Reanalyze (constant BPM) 
* Reanalyze (variable BPM) 

* Analyze
* Reanalyze
[✓] variable BPM 

default [✓] is from the preferences option

fwcd commented:
I think @ronso0's

- Analyze
- Reanalyze
- Reanalyze with [opposite of setting]

strikes a good balance. The checkbox one is intriguing too, but I like the above one because it saves a click (since, in my experience, a user that wants to reanalyze is likely to pick the [opposite of setting] option, especially if they already performed a batch analysis using the default setting).

fwcd commented:
Since we'd need to update the menus whenever the setting changes, is there an elegant way to hook into/listen for settings changes? Or would we need to add a new signal e.g. in `beatdetectionsettings`?

daschuer commented:
You may pipe the userSettings pointer into the class and recreate the menu just before showing.

The other option is to backup the preferences option with a ControlObject and consume the value change signal.  

I think you will find templaes for both in the Mixxx code.

fwcd commented:
Hm, I'm almost inclined to do that in a separate PR to keep this one focused if it requires a lot of wiring to make the menu dynamic.

daschuer commented:
That would be OK for me.

fwcd committed:
WTrackMenu: Hide the redundant reanalysis action

fwcd commented:
Ah thanks, I wasn't aware of the `updateMenus` method. The menu should now automatically update accordingly.

daschuer commented:
We now have basically: 
`std::optional<std::not_null<std::shared_ptr<Track>>>`

You introduce std::not_null to justify the use of std::optional and this comes with a lot of surprising magic and is slowing down the process. 

I assume that you are coming from languages where the std::optional equivalent is the only solution to express a null value. This is not the case in C++ where a pointer is already an "optional". Also the used std::shared_ptr has the "optional semantic" https://en.cppreference.com/w/cpp/memory/shared_ptr/operator_bool. 
So for a C++ developer is unusual to not use this feature, but use a not free workaround instead. Also in Qt many classes are optional. For instance https://doc.qt.io/qt-6/qvariant.html#isNull. In Mixxx we also use this pattern Bpm::isValid();

This is a well established pattern in C++ and I strongly recommend to use it here. 

Let's move the Track not_null idea to a separate PR. I think we should create our own not_null track class that fixes the performance issues.  
 
Apart form that, this branch works good. Thank you.

fwcd commented:
I'd say it's more like

```cpp
std::optional<std::pair<
  gsl::not_null<std::shared_ptr<Track>>,
  AnalyzerTrack::Options
>>
```

which is not isomorphic to the type you described:

```cpp
std::pair<
  std::shared_ptr<Track>,
  AnalyzerTrack::Options
>
```

So I'd argue that even though the difference is subtle, they do have different semantics and I would also argue that the upper one is the more correct one (since it doesn't really make sense to specify options on a `null` track).

> a lot of surprising magic and is slowing down the process.

I really don't think this is a fair assessment of the situation. `std::optional` is safer, more explicit, better signals intent and forces the programmer to stop for a moment and think about optionality. How is a nullable pointer where the programmer may or may not have intended to allow `null` better?

> So for a C++ developer is unusual to not use this feature, but use a not free workaround instead. Also in Qt many classes are optional. For instance https://doc.qt.io/qt-6/qvariant.html#isNull. In Mixxx we also use this pattern Bpm::isValid();

I agree, it's a common pattern to write custom `isValid`-style methods. But this is an "appeal to tradition" argument: If we have the opportunity to modernize the codebase here in the face of a _new_ feature, why not do so directly instead of implementing the legacy pattern and then refactoring later on?

> I assume that you are coming from languages where the std::optional equivalent is the only solution to express a null value.

I have worked in a bunch of languages, including ones that implicitly permit `null` everywhere (Java, Python etc.) and languages that enforce strongly-typed optionals (Rust, Haskell, Kotlin, Swift, ...). Encoding invariants at the type-level is not only a general trend in modern, statically-typed languages, even C++ seems to move into the direction. It's the reason why we use smart pointers instead of managing memory manually using `new` and `delete` everywhere (aside from situations where tiny differences in memory usage are actually critical, but this isn't really the case here), etc.

> fixes the performance issues.

We already do a null check, so there's no performance overhead. `gls::not_null` also has zero overhead as mentioned in the docs, `std::optional` maybe a byte. This is a minuscule space cost per `AnalyzerThread` that's not really worth bikeshedding over.

> Let's move the Track not_null idea to a separate PR.

This only applies to `AnalyzerTrack` not all other `TrackPointer`s. I don't think the general pattern of using `null` `TrackPointer`s is bad, just that it makes sense for this particular use case to bundle a not-null track with it's options and _then_ add optionality on top.

uklotzde commented:
@fwcd I appreciate your efforts to modernize the code base and fully agree with your arguments about ["encoding invariants at the type-level"](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).

ywwg commented:
for what it's worth, using nullptr as an optional type has several shortcomings -- code can crash if it doesn't handle null correctly, and variables can end up being null even if they weren't intended to be so.  std::optional is a good feature to make this use explicit and does not impose a lot of computational overhead.  So I am in favor of using a more modern approach than using pointers-and-null

daschuer commented:
I sounds like you did not get my point correctly.
I am not against using std::optional here.

The remaining issue in this case is wrapping ``std::not_null<std::shared_ptr<Track>>``
Every access off the pointer is a deep copy of the shared pointer including all book keeping and memory fences for syncing. 

That needs to be solved! 

I and I have already recomended to move that out of this PR. Just to continue here ...

daschuer commented:
@ywwg std::optional suffers exactly the same issue as plain pointer. If you access it without a check Mixxx crashes.

daschuer commented:
Summary: I see two possible solutions: 
* Implement an TrackPointerNotNull where operator-> and friends are raw pointer access. Than use it here warped in an std::optional 
* Add the optional semantic to AnalyzerTrack, and reuse the null TrackPointer to express this.

fwcd commented:
> Implement an TrackPointerNotNull where operator-> and friends are raw pointer access. Than use it here warped in an std::optional

This is pretty much what we are doing right now. Since `gsl::not_null` is a zero-overhead wrapper, I don't see how a custom class would be necessary instead of just wrapping the `shared_ptr`. Any refcounting overhead here would already be present in the current Mixxx and optimizing this would IMO indeed be out of scope for this PR.

daschuer commented:
> Since gsl::not_null is a zero-overhead wrapper ... 

I have just realized that since:  
https://github.com/microsoft/GSL/commit/4a4bb3c13a6c782b9207bba9ece72b123ff76431

This is actually the case. My concerns where regarding the previous version, which has the described overhead.

daschuer commented:
Wait, I have misread the code 
`std::is_copy_constructible<std::shared_ptr<T>>::value` 
is true this mean we do a deep copy, which the opposite of zero overhead. 

The changes where introduced to support std::unique_ptr() 

So my conclusion is still stands: 
The introduced overhead is not worth the benefits we have by using std::optional, because the alternative traditional solution offers the same safety and semantics and is much more performant. 

We may add a wrapper header that defines `is_copy_constructible` as `is_pod`. 
Or inherit from std::shared_ptr a new one with deleted copy constructor. 
We may just copy and edit the not_null template ...

Ideas?

fwcd commented:
The docs explicitly state that there's no overhead: https://github.com/microsoft/GSL/blob/1683d87a3f45dff1817e858dbaf006e5ff84e784/include/gsl/pointers#L74 (this is a permalink, but points to the current main branch). But even if there were, I think vendoring our own version is not a good idea, for reasons mentioned in my previous comments:

> going with battle-tested (and optimized) standard library types like shared_ptr is IMO usually better than rolling our own with all of the pitfalls that writing your own wrappers comes with (e.g. making sure all copy/move constructors/assignments are implemented correctly, just to name one example). Another advantage is that using types from the standard library makes the codebase less obscure and more accessible to new contributors, since they are likely to be familiar with these types.

This is really not a case worth micro-optimizing. Tracks added to the analyzer queue change maybe once every few seconds and `std::optional` adds so little overhead that, unless we have evidence that it actually is a space/performance issue (e.g. by profiling it), I think the tradeoff in productivity/readability is simply not worth it.

daschuer commented:
I have found a related bug https://github.com/microsoft/GSL/issues/550

> The docs explicitly state that there's no overhead: https://github.com/microsoft/GSL/blob/1683d87a3f45dff1817e858dbaf006e5ff84e784/include/gsl/pointers#L74

This is regarding size. My concerns are regarding runtime.  

> With "deep copy" I assume you mean "copying the pointer to a track and maybe updating its reference count"?

Yes, it counts always fore and back and may empty the CPU pipeline because of memory fences. 

> This is really not a case worth micro-optimizing.

Look, we compare here the traditional solution with a solution that aims to give a better semantic. Both are ways of optimization. I am happy with the later if it actually has close to zero overhead as assumed, but since this is not yet the case, let's go for the traditional solution. 

It would be also a good solution if the we remove the deep copy somehow.

fwcd commented:
Hm, that's a fair point, I wasn't aware that `not_null<shared_ptr<...>>` actually performed a reference increment/decrement on every access. I still think that it wouldn't really be worth optimizing, since infrequent reference copies are comparatively cheap. But maybe we can find a solution that works for all of us.

What do you think about adding an assertion or check to `AnalyzerTrack`'s constructor that the passed `TrackPointer` is not null and then perhaps a doc comment that `AnalyzerTrack` always wraps a not-null pointer? Since the wrapped `TrackPointer` field is not accessible from the outside, this should guarantee that the invariant (the pointer not being null) is always upheld. To avoid unnecessary reference copies we could make `getTrack` return the `shared_ptr` by `const` reference and anyone interested in retaining the `TrackPointer` could then make a reference copy themselves.

daschuer commented:
Good idea. That works for me.

fwcd committed:
AnalyzerTrack: Assert non-nullness and return track by ref

As per the discussion: https://github.com/mixxxdj/mixxx/pull/10931#issuecomment-1267275389

fwcd commented:
Done!

daschuer commented on src/analyzer/analyzerthread.cpp:
```diff
@@ -126,16 +126,16 @@ void AnalyzerThread::doRun() {
     while (awaitWorkItemsFetched()) {
-        DEBUG_ASSERT(m_currentTrack);
-        kLogger.debug() << "Analyzing" << m_currentTrack->getFileInfo();
+        DEBUG_ASSERT(m_currentTrack.has_value());
```
```suggestion
        VERIFY_OR_DEBUG_ASSERT(m_currentTrack.has_value()) {
            continue; 
        }
```
because the has_value check is out of sight, and we don't want to crash in a release version.

fwcd replied:
As above, I don't think any logic changed here, since `has_value()` doesn't have any side effects, I don't think that should be necessary (assuming of course, there wasn't some crash hiding in here before).

daschuer commented on src/analyzer/analyzerthread.cpp:
```diff
@@ -187,11 +187,11 @@ void AnalyzerThread::doRun() {
-bool AnalyzerThread::submitNextTrack(TrackPointer nextTrack) {
-    DEBUG_ASSERT(nextTrack);
+bool AnalyzerThread::submitNextTrack(const AnalyzerTrack& nextTrack) {
+    DEBUG_ASSERT(nextTrack.getTrack());
```
This can be removed, right? Juchhu :-)

fwcd replied:
Yes, good catch!

daschuer commented on src/analyzer/analyzertrack.h:
```diff
@@ -0,0 +1,28 @@
+
+  private:
+    /// The (not-null) track to be analyzed.
+    TrackPointer track;
```
```suggestion
    TrackPointer m_track;
```

daschuer commented on src/analyzer/analyzertrack.h:
```diff
@@ -0,0 +1,28 @@
+    /// The (not-null) track to be analyzed.
+    TrackPointer track;
+    /// The additional options.
+    Options options;
```
```suggestion
    Options m_options;
```

daschuer reviewed (changes requested):
Thank you. I have just added some code style comments.

fwcd commented on src/analyzer/analyzerscheduledtrack.cpp:
```diff
@@ -0,0 +1,16 @@
+        : trackId(trackId), options(options) {
+}
+
+const TrackId AnalyzerScheduledTrack::getTrackId() const {
```
I did intend to return by value here since I understood `TrackId` to be a small, trivially copyable type. You are right that `const` doesn't make much sense in that case. Do you think it would be better to return by ref here?

daschuer replied:
I think `sizeof(TrackId)` = 4 so that can be safely returned by value. On the other hand a receiving copy constructor takes always a const reference, so that this should be the preferred method of returning member variables. 

I would go for 
```suggestion
const TrackId& AnalyzerScheduledTrack::getTrackId() const {
```

fwcd commented on src/analyzer/analyzerthread.cpp:
```diff
@@ -341,18 +340,18 @@ void AnalyzerThread::emitBusyProgress(AnalyzerProgress busyProgress) {
 
 void AnalyzerThread::emitDoneProgress(AnalyzerProgress doneProgress) {
-    DEBUG_ASSERT(m_currentTrack);
+    DEBUG_ASSERT(m_currentTrack.has_value());
```
I'm a bit confused, this doesn't really change any logic (apart from migrating the null check to an optional check), wouldn't this have crashed Mixxx before if that were the case?

daschuer replied:
Yes, right.

fwcd commented on src/analyzer/analyzerscheduledtrack.cpp:
```diff
@@ -0,0 +1,16 @@
+    return trackId;
+}
+
+const AnalyzerTrack::Options AnalyzerScheduledTrack::getOptions() const {
```
I have no strong opinion on whether by-ref or by-value is better here, it's not a large type currently (2 bytes), but since it might get bigger in the future returning by ref might make sense to future-proof this against large copies (though it's probably not accessed that often anyways).

fwcd commented on src/analyzer/analyzerscheduledtrack.h:
```diff
@@ -0,0 +1,26 @@
+
+  private:
+    /// The id of the track to be analyzed.
+    TrackId trackId;
```
No opinion on Hungarian notation either, I have seen classes that use it and some that don't in the codebase. But for consistency, it's probably reasonable to use it here too.

daschuer replied:
This is part of our code style: 
https://github.com/mixxxdj/mixxx/wiki/Coding-Guidelines#classes

fwcd committed:
Analyzer: Apply style suggestions as per PR review

See https://github.com/mixxxdj/mixxx/pull/10931#pullrequestreview-1128977177

daschuer reviewed (approved):
Thank you for this nice improvement. LGTM
